src/Google/Provider.php

```diff
@@ -2,13 +2,188 @@
 
 namespace SocialiteProviders\Google;
 
-use Laravel\Socialite\Two\GoogleProvider;
-use SocialiteProviders\Manager\ConfigTrait;
-use SocialiteProviders\Manager\Contracts\OAuth2\ProviderInterface;
+use Illuminate\Http\Request;
+use Illuminate\Support\Arr;
+use SocialiteProviders\Manager\OAuth2\AbstractProvider;
 
-class Provider extends GoogleProvider implements ProviderInterface
+class Provider extends AbstractProvider
 {
-    use ConfigTrait;
+    /**
+     * The authenticated user instance.
+     *
+     * @var GoogleUser
+     */
+    protected $user;
 
-    public const IDENTIFIER = 'GOOGLE';
+    /**
+     * The current request instance.
+     *
+     * @var \Illuminate\Http\Request
+     */
+    protected $request;
+
+    /**
+     * The client ID.
+     *
+     * @var string
+     */
+    protected $clientId;
+
+    /**
+     * The decoded JWT payload.
+     *
+     * @var array
+     */
+    protected $payload;
+
+    /**
+     * Create a new Google provider instance.
+     *
+     * @param \Illuminate\Http\Request $request
+     * @param string                   $clientId
+     */
+    public function __construct(Request $request, $clientId)
+    {
+        $this->request = $request;
+        $this->clientId = $clientId;
+    }
+
+    /**
+     * Redirect the user to the authentication page for the provider.
+     *
+     * @throws \Exception
+     */
+    public function redirect()
+    {
+        throw new \RuntimeException('Redirect is deprecated for the new Google Auth, see https://developers.google.com/identity/gsi/web/guides/migration#html_and_javascript');
+    }
+
+    /**
+     * Get the User instance for the authenticated user.
+     *
+     * @throws InvalidJwtException
+     *
+     * @return GoogleUser
+     */
+    public function user()
+    {
+        if ($this->user) {
+            return $this->user;
+        }
+        $jwt = $this->request->credential;
```

Comment:
Usually JWT libraries already handle this for you and raise the right exception with the right message

Might be worth leveraging their advanced control features

Reply:
I've tried to achieve the functionality without using any external library.

Reply:
It is fine using one 👍 
For Apple uses lcobucci/jwt and Azure, Eve and Xero use firebase/php-jwt